0006-ffmul-revised.md

```diff
@@ -0,0 +1,1932 @@
+
+$$
+\begin{aligned}
+limbs_{max} &= \lfloor cells/4  \rfloor \\
+      &= \lfloor 14/4 \rfloor \\
+      &= 3 \\
+\end{aligned}
+$$
+
+Thus, the maximum number of limbs possible in a single gate configuration is 3.
+
+Using $limbs_{max}=3$ and $t=264$ that covers our use cases (see the previous section), we are finally able to derive our limbs size
+
+$$
+\begin{aligned}
+\ell &= \frac{t}{limbs_{max}} \\
+&= 264/3 \\
+&= 88
+\end{aligned}
+$$
+
+Therefore, our limb configuration is:
+
+  * Limb size $\ell = 88$ bits
+  * Number of limbs is $3$
+
+## Appendix C: Borrows
+
+When we constrain $a \cdot b - q \cdot f = r \mod 2^t$ we want to be as efficient as possible.
+
+Observe that the expansion of $a \cdot b - q \cdot f$ into limbs would also have subtractions between limbs, requiring our constraints to account for borrowing.  Dealing with this would create undesired complexity and increase the degree of our constraints.
+
+In order to avoid the possibility of subtractions we instead use $a \cdot b + q \cdot f'$ where
+
+$$
+\begin{aligned}
+f' &= -f \mod 2^t \\
+   &= 2^t - f
+\end{aligned}
+$$
+
+The negated modulus $f'$ becomes part of our gate coefficients and is not constrained because it is publicly auditable.
+
+Using the substitution of the negated modulus, we now must constrain $a \cdot b + q \cdot f' = r \mod 2^t$.
+
+> Observe that $f' < 2^t$ since $f < 2^t$ and that $f' > f$ when $f < 2^{t - 1}$.
+
+### Subtractions
+
+When unconstrained, computing $u_0 = p_0 + 2^{\ell} \cdot p_{10} - (r_0 + 2^{\ell} \cdot r_1)$ could require borrowing.  Fortunately, we have the constraint that the $2\ell$ least significant bits of $u_0$ are `0` (i.e. $u_0 = 2^{2\ell} \cdot v_0$), which means borrowing cannot happen.
+
+Borrowing is prevented because when the the $2\ell$ least significant bits of the result are `0` it is not possible for the minuend to be less than the subtrahend.  We can prove this by contradiction.
+
+Let
+
+* $x = p_0 + 2^{\ell} \cdot p_{10}$
+* $y = r_0 + 2^{\ell} \cdot r_1$
+* the $2\ell$ least significant bits of $x - y$ be `0`
+
+Suppose that borrowing occurs, that is, that $x < y$.  Recall that the length of $x$ is at most $2\ell + 2$ bits.  Therefore, since $x < y$ the top two bits of $x$ must be zero and so we have
+
+$$
+x - y = x_{2\ell} - y,
+$$
+
+where $x_{2\ell}$ denotes the $2\ell$ least significant bits of $x$.
+
+Recall also that the length of $y$ is $2\ell$ bits.  We know this because limbs of the result are each constrained to be in $[0, 2^{\ell})$.  So the result of this subtraction is $2\ell$ bits.   Since the $2\ell$ least significant bits of the subtraction are `0` this means that
+
+$$
+\begin{aligned}
+x - y  &  = 0 \\
+x &= y,
+\end{aligned}
+$$
+
+which is a contradiction with $x < y$.
+
+## Appendix D: Gate constraints
```

Comment:
This whole section feels redundant with the main body. Would be better to replace this whole section with just the diagrams, and refer to them from the relevant points.

Reply:
Yeah, that's true. I will then keep diagrams only.

Reply:
I found some diagrams and equations to go well with a bit of text. I reduced the amount of sections to only keep the ones containing actual diagrams and "novel" content.